File: Dev/C++/Mascaret/src/VEHA/Behavior/Common/Trigger.cpp
#include "VEHA/Behavior/Common/Trigger.h"
#include "VEHA/Behavior/Common/Event.h"
#include "VEHA/Behavior/Common/Signal.h"
#include "VEHA/Behavior/StateMachine/SignalEvent.h"
#include "VEHA/Behavior/StateMachine/ChangeEvent.h"
#include "VEHA/Behavior/StateMachine/TimeEvent.h"
#include "VEHA/Kernel/Expression.h"

namespace VEHA
{

Trigger::Trigger(const string& name)
:Element(name)
{
}

Trigger::~Trigger()
{
}

bool Trigger::equals(shared_ptr<Trigger> trigger)
{
	if (_event->getType() != trigger->_event->getType()) return false;
	if ( _event->getType() == "SignalEvent")
	{
//cerr << (shared_dynamic_cast<SignalEvent>  (_event))->getSignal()->getName();
//cerr << " == ";
//cerr << (shared_dynamic_cast<SignalEvent>  (trigger->_event))->getSignal()->getName();
//cerr << endl; 
                //cerr << (shared_dynamic_cast<SignalEvent>  (_event))->getSignal() << endl;
		if ((shared_dynamic_cast<SignalEvent>  (_event))->getSignalClass()->getName() == (shared_dynamic_cast<SignalEvent>  (trigger->_event))->getSignal()->getClassifier()->getName()) return true;
		else return false;
	}
	else if (_event->getType() == "ChangeEvent")
	{
		shared_ptr<Expression> exp1 = shared_dynamic_cast<Expression>((shared_dynamic_cast<ChangeEvent>(_event))->getChangeExpression());
		shared_ptr<Expression> exp2 = shared_dynamic_cast<Expression>((shared_dynamic_cast<ChangeEvent>(trigger->_event))->getChangeExpression());
		if ( exp1->getExpression() == exp2->getExpression() )
			return true;
		else return false;
	}
	else if (_event->getType() == "TimeEvent")
	{
		shared_ptr<TimeEvent> te1 = shared_dynamic_cast<TimeEvent>(_event);
		shared_ptr<TimeEvent> te2 = shared_dynamic_cast<TimeEvent>(trigger->_event);
		shared_ptr<LiteralReal> value1 = shared_dynamic_cast<LiteralReal>(te1->getWhen());
		double timeValue1 = value1->getValue();
		shared_ptr<LiteralReal> value2 = shared_dynamic_cast<LiteralReal>(te2->getWhen());
		double timeValue2 = value2->getValue();

		if ( timeValue1 == timeValue2 )
			return true;
		else return false;
	}
	else
		return false;
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Behavior/Common/Trigger.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Mascaret
{
    public class Trigger : Element
    {

        private MascaretEvent mEvent;
        public MascaretEvent MEvent
        {
            get { return mEvent; }
            set { mEvent = value; }
        }

        public Trigger(string name)
            : base(name)
        {
        }

        public bool equals(Trigger trigger)
        {
            if (mEvent.Type != trigger.mEvent.Type)
                return false;
            if (mEvent.Type == "SignalEvent")
            {
               // StreamWriter file = MascaretApplication.Instance.logfile;
                if (mEvent as SignalEvent != null && ((SignalEvent)mEvent).name == ((SignalEvent)trigger.mEvent).Signal.Classifier.name)
                {
                    return true;
                }
                else
                    return false;
            }
            else if (mEvent.Type == "ChangeEvent")
            {
                Expression exp1 = ((Expression)((ChangeEvent)mEvent).ChangeExpression);
                Expression exp2 = ((Expression)((ChangeEvent)trigger.MEvent).ChangeExpression);
                if (exp1.ExpressionValue == exp2.ExpressionValue)
                    return true;
                else return false;
            }
            else if (mEvent.Type == "TimeEvent")
            {
                LiteralReal value1 = (LiteralReal)((TimeEvent)mEvent).When;
                LiteralReal value2 = (LiteralReal)((TimeEvent)trigger.MEvent).When;

                if (value1.RValue == value2.RValue)
                    return true;
                else return false;
            }
            else
                return false;
        }
    }
}

